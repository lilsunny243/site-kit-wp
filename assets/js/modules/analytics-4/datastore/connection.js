/**
 * `modules/analytics-4` connection data store
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_MODULES } from '../../../googlesitekit/modules/datastore/constants';
import {
	isValidMeasurementID,
	isValidPropertySelection,
	isValidWebDataStreamID,
} from '../utils/validation';
import { MODULES_ANALYTICS_4 } from './constants';
const { createRegistrySelector } = Data;

export const selectors = {
	/**
	 * Checks if GA4 is connected, pending module data refresh.
	 */
	isFreshlyConnected: createRegistrySelector( ( select ) => ( state ) => {
		// If the module is already connected, it isn't "freshly" connected.
		if ( select( CORE_MODULES ).isModuleConnected( 'analytics-4' ) ) {
			return false;
		}
		// Wait for settings to be available + ensure settings resolver is invoked.
		if (
			undefined === select( MODULES_ANALYTICS_4 ).getSettings() ||
			! state.savedSettings
		) {
			return undefined;
		}
		// "Calculate" the connected state based on saved settings.
		return (
			isValidPropertySelection( state.savedSettings.propertyID ) &&
			isValidWebDataStreamID( state.savedSettings.webDataStreamID ) &&
			isValidMeasurementID( state.savedSettings.measurementID )
		);
	} ),
};

const store = { selectors };

export default store;
