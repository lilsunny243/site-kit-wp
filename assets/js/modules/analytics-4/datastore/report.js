/**
 * `modules/analytics-4` data store: report.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * External dependencies
 */
import invariant from 'invariant';
import { isPlainObject } from 'lodash';

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
import { CORE_SITE } from '../../../googlesitekit/datastore/site/constants';
import { DATE_RANGE_OFFSET, MODULES_ANALYTICS_4 } from './constants';
import { DAY_IN_SECONDS, stringifyObject } from '../../../util';
import { isValidDateRange } from '../../../util/report-validation';
import {
	normalizeReportOptions,
	isValidDimensionFilters,
	isValidDimensions,
	isValidMetrics,
	isValidOrders,
	isZeroReport,
} from '../utils';
import { createGatheringDataStore } from '../../../googlesitekit/modules/create-gathering-data-store';
const { createRegistrySelector } = Data;

/**
 * Returns report args for a sample report.
 *
 * @since 1.107.0
 *
 * @param {Function} select The select function of the registry.
 * @return {Object} Report args.
 */
const getSampleReportArgs = ( select ) => {
	const { startDate, endDate } = select( CORE_USER ).getDateRangeDates( {
		offsetDays: DATE_RANGE_OFFSET,
	} );

	const args = {
		dimensions: [ 'date' ],
		metrics: [ { name: 'totalUsers' } ],
		startDate,
		endDate,
	};

	const url = select( CORE_SITE ).getCurrentEntityURL();
	if ( url ) {
		args.url = url;
	}

	return args;
};

const fetchGetReportStore = createFetchStore( {
	baseName: 'getReport',
	controlCallback: ( { options } ) => {
		return API.get(
			'modules',
			'analytics-4',
			'report',
			normalizeReportOptions( options )
		);
	},
	reducerCallback: ( state, report, { options } ) => {
		return {
			...state,
			reports: {
				...state.reports,
				[ stringifyObject( options ) ]: report,
			},
		};
	},
	argsToParams: ( options ) => {
		return { options };
	},
	validateParams: ( { options } = {} ) => {
		invariant(
			isPlainObject( options ),
			'options for Analytics 4 report must be an object.'
		);
		invariant(
			isValidDateRange( options ),
			'Either date range or start/end dates must be provided for Analytics 4 report.'
		);

		const { metrics, dimensions, dimensionFilters, orderby } =
			normalizeReportOptions( options );

		invariant(
			metrics.length,
			'Requests must specify at least one metric for an Analytics 4 report.'
		);
		invariant(
			isValidMetrics( metrics ),
			'metrics for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects, or a mix of strings and objects. Objects must have a "name" property. Metric names must match the expression ^[a-zA-Z0-9_]+$.'
		);

		if ( dimensions ) {
			invariant(
				isValidDimensions( dimensions ),
				'dimensions for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects, or a mix of strings and objects. Objects must have a "name" property.'
			);
		}

		if ( dimensionFilters ) {
			invariant(
				isValidDimensionFilters( dimensionFilters ),
				'dimensionFilters for an Analytics 4 report must be a map of dimension names as keys and dimension values as values.'
			);
		}

		if ( orderby ) {
			invariant(
				isValidOrders( orderby ),
				'orderby for an Analytics 4 report must be an array of OrderBy objects where each object should have either a "metric" or "dimension" property, and an optional "desc" property.'
			);
		}
	},
} );

const gatheringDataStore = createGatheringDataStore( 'analytics-4', {
	storeName: MODULES_ANALYTICS_4,
	dataAvailable:
		global._googlesitekitModulesData?.[ 'data_available_analytics-4' ],
	selectDataAvailability: createRegistrySelector( ( select ) => () => {
		// Disable reason: This needs to be run here in order for the report to be resolved.
		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
		const hasZeroData = select( MODULES_ANALYTICS_4 ).hasZeroData();

		const args = getSampleReportArgs( select );

		const hasResolvedReport = select(
			MODULES_ANALYTICS_4
		).hasFinishedResolution( 'getReport', [ args ] );

		if ( ! hasResolvedReport || hasZeroData === undefined ) {
			return undefined;
		}

		const hasReportError = select(
			MODULES_ANALYTICS_4
		).getErrorForSelector( 'getReport', [ args ] );

		// If there is an error, return `null` since we don't know if there is data or not.
		if ( hasReportError ) {
			return null;
		}

		if ( hasZeroData === false ) {
			return true;
		}

		const isAuthenticated = select( CORE_USER ).isAuthenticated();

		if ( isAuthenticated === undefined ) {
			return undefined;
		}
		if ( ! isAuthenticated ) {
			return false;
		}

		const propertyID = select( MODULES_ANALYTICS_4 ).getPropertyID();

		if ( propertyID === undefined ) {
			return undefined;
		}

		const property =
			select( MODULES_ANALYTICS_4 ).getProperty( propertyID );

		if ( property === undefined ) {
			return undefined;
		}

		const createTime = new Date( property.createTime ).getTime();

		// If the property was created within the last three days and has no data, assume it's still gathering data.
		if ( createTime > Date.now() - DAY_IN_SECONDS * 3 * 1000 ) {
			return false;
		}

		return true;
	} ),
} );

const baseInitialState = {
	reports: {},
};

const baseResolvers = {
	*getReport( options = {} ) {
		const registry = yield Data.commonActions.getRegistry();
		const existingReport = registry
			.select( MODULES_ANALYTICS_4 )
			.getReport( options );

		// If there is already a report loaded in state, consider it fulfilled
		// and don't make an API request.
		if ( existingReport ) {
			return;
		}

		yield fetchGetReportStore.actions.fetchGetReport( options );
	},
};

const baseSelectors = {
	/**
	 * Gets an Analytics report for the given options.
	 *
	 * @since 1.94.0
	 *
	 * @param {Object}         state                      Data store's state.
	 * @param {Object}         options                    Options for generating the report.
	 * @param {string}         options.startDate          Required. Start date to query report data for as YYYY-mm-dd.
	 * @param {string}         options.endDate            Required. End date to query report data for as YYYY-mm-dd.
	 * @param {Array.<string>} options.metrics            Required. List of metrics to query.
	 * @param {string}         [options.compareStartDate] Optional. Start date to compare report data for as YYYY-mm-dd.
	 * @param {string}         [options.compareEndDate]   Optional. End date to compare report data for as YYYY-mm-dd.
	 * @param {Array.<string>} [options.dimensions]       Optional. List of dimensions to group results by. Default an empty array.
	 * @param {Object}         [options.dimensionFilters] Optional. Map of dimension filters for filtering options on a dimension. Default an empty object.
	 * @param {Array.<Object>} [options.orderby]          Optional. An order definition object, or a list of order definition objects, each one containing 'fieldName' and 'sortOrder'. 'sortOrder' must be either 'ASCENDING' or 'DESCENDING'. Default empty array.
	 * @param {string}         [options.url]              Optional. URL to get a report for only this URL. Default an empty string.
	 * @param {number}         [options.limit]            Optional. Maximum number of entries to return. Default 1000.
	 * @return {(Array.<Object>|undefined)} An Analytics report; `undefined` if not loaded.
	 */
	getReport( state, options ) {
		const { reports } = state;

		return reports[ stringifyObject( options ) ];
	},

	/**
	 * Gets a Page title to URL map for the given options.
	 *
	 * @since 1.96.0
	 *
	 * @param {Object} state             Data store's state.
	 * @param {Object} report            A report from the getReport selector containing pagePaths.
	 * @param {Object} options           Options for generating the report.
	 * @param {string} options.startDate Required, start date to query report data for as YYYY-mm-dd.
	 * @param {string} options.endDate   Required, end date to query report data for as YYYY-mm-dd.
	 * @return {(Object|undefined)} A map with url as the key and page title as the value. `undefined` if not loaded.
	 */
	getPageTitles: createRegistrySelector(
		( select ) =>
			( state, report, { startDate, endDate } = {} ) => {
				if ( ! isPlainObject( report ) ) {
					return;
				}

				const pagePaths = [];
				const REQUEST_MULTIPLIER = 5;

				/*
				 * Iterate over the report rows, finding the dimension containing the
				 * pagePath value which we add to the array of pagePaths.
				 */
				const { dimensionHeaders, rows } = report;
				if (
					Array.isArray( dimensionHeaders ) &&
					Array.isArray( rows )
				) {
					const pagePathIndex = dimensionHeaders.findIndex(
						( { name } ) => name === 'pagePath'
					);
					rows.forEach( ( { dimensionValues } ) => {
						if (
							! pagePaths.includes(
								dimensionValues[ pagePathIndex ].value
							)
						) {
							pagePaths.push(
								dimensionValues[ pagePathIndex ].value
							);
						}
					} );
				}

				const urlTitleMap = {};
				if ( ! pagePaths.length ) {
					return urlTitleMap;
				}

				const options = {
					startDate,
					endDate,
					dimensions: [ 'pagePath', 'pageTitle' ],
					dimensionFilters: { pagePath: pagePaths.sort() },
					metrics: [ { name: 'screenPageViews' } ],
					orderby: [
						{
							metric: { metricName: 'screenPageViews' },
							desc: true,
						},
					],
					limit: REQUEST_MULTIPLIER * pagePaths.length,
				};

				const pageTitlesReport =
					select( MODULES_ANALYTICS_4 ).getReport( options );
				if ( undefined === pageTitlesReport ) {
					return;
				}

				( pageTitlesReport?.rows || [] ).forEach(
					( { dimensionValues } ) => {
						if ( ! urlTitleMap[ dimensionValues[ 0 ].value ] ) {
							// key is the url, value is the page title.
							urlTitleMap[ dimensionValues[ 0 ].value ] =
								dimensionValues[ 1 ].value;
						}
					}
				);

				pagePaths.forEach( ( pagePath ) => {
					if ( ! urlTitleMap[ pagePath ] ) {
						// If we don't have a title for the pagePath, we use '(unknown)'.
						urlTitleMap[ pagePath ] = __(
							'(unknown)',
							'google-site-kit'
						);
					}
				} );

				return urlTitleMap;
			}
	),

	/**
	 * Determines whether Analytics 4 has zero data or not.
	 *
	 * @since 1.95.0
	 * @since 1.107.0 Returns `true` if the report request has an error to be consistent with `hasZeroData` selectors of other modules.
	 *
	 * @return {boolean|undefined} Returns `true` if the report is zero, otherwise `false`. Returns `undefined` while resolving.
	 */
	hasZeroData: createRegistrySelector( ( select ) => () => {
		const args = getSampleReportArgs( select );

		// Disable reason: select needs to be called here or it will never run.
		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
		const report = select( MODULES_ANALYTICS_4 ).getReport( args );
		const hasResolvedReport = select(
			MODULES_ANALYTICS_4
		).hasFinishedResolution( 'getReport', [ args ] );

		if ( ! hasResolvedReport ) {
			return undefined;
		}

		const hasReportError = select(
			MODULES_ANALYTICS_4
		).getErrorForSelector( 'getReport', [ args ] );

		// If there is an error, we assume it's a zero report.
		if ( hasReportError ) {
			return true;
		}

		return isZeroReport( report );
	} ),
};

const store = Data.combineStores( fetchGetReportStore, gatheringDataStore, {
	initialState: baseInitialState,
	resolvers: baseResolvers,
	selectors: baseSelectors,
} );

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
