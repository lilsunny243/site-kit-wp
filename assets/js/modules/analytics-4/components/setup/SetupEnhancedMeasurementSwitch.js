/**
 * SetupEnhancedMeasurementSwitch component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useEffect, useRef } from '@wordpress/element';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { MODULES_ANALYTICS } from '../../../analytics/datastore/constants';
import {
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import EnhancedMeasurementSwitch from '../common/EnhancedMeasurementSwitch';
const { useSelect, useDispatch } = Data;

export default function SetupEnhancedMeasurementSwitch() {
	const accountID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS ).getAccountID()
	);

	const propertyID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getPropertyID()
	);

	const webDataStreamID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
	);

	const isEnhancedMeasurementEnabled = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue(
			ENHANCED_MEASUREMENT_FORM,
			ENHANCED_MEASUREMENT_ENABLED
		)
	);

	const { setValues } = useDispatch( CORE_FORMS );

	// If `isEnhancedMeasurementEnabled` is already defined in the first render, it means we're rendering this component
	// after the user has actively selected the setting, in which case we don't want to override the setting to `true` in
	// the `useEffect()` unless the account, property, or web data stream selection is subsequently changed.
	const skipEffect = useRef( isEnhancedMeasurementEnabled !== undefined );

	useEffect( () => {
		if ( skipEffect.current ) {
			skipEffect.current = false;
			return;
		}

		setValues( ENHANCED_MEASUREMENT_FORM, {
			[ ENHANCED_MEASUREMENT_ENABLED ]: true,
		} );
	}, [ accountID, propertyID, webDataStreamID, setValues ] );

	return <EnhancedMeasurementSwitch />;
}
