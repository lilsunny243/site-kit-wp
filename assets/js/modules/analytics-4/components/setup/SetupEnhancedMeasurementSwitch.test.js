/**
 * SetupEnhancedMeasurementSwitch tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	act,
	createTestRegistry,
	render,
} from '../../../../../../tests/js/test-utils';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { MODULES_ANALYTICS } from '../../../analytics/datastore/constants';
import {
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	MODULES_ANALYTICS_4,
	PROPERTY_CREATE,
	WEBDATASTREAM_CREATE,
} from '../../datastore/constants';
import SetupEnhancedMeasurementSwitch from './SetupEnhancedMeasurementSwitch';

describe( 'SetupEnhancedMeasurementSwitch', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();

		registry.dispatch( MODULES_ANALYTICS ).receiveGetSettings( {
			accountID: '1000',
		} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			propertyID: '2000',
			webDataStreamID: '3000',
		} );
	} );

	it( 'should render with the switch defaulting to the on position', () => {
		const { container, getByLabelText } = render(
			<SetupEnhancedMeasurementSwitch />,
			{
				registry,
			}
		);

		expect( container ).toMatchSnapshot();

		const switchControl = getByLabelText( 'Enable enhanced measurement' );

		expect( switchControl ).toBeChecked();
	} );

	it.each( [
		[ 'propertyID', PROPERTY_CREATE ],
		[ 'webDataStreamID', WEBDATASTREAM_CREATE ],
	] )(
		'should not default the switch to the on position when the %s is initially %s and `isEnhancedMeasurementEnabled` is already `false`',
		( settingName, settingCreate ) => {
			registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
				[ settingName ]: settingCreate,
			} );

			registry
				.dispatch( CORE_FORMS )
				.setValues( ENHANCED_MEASUREMENT_FORM, {
					[ ENHANCED_MEASUREMENT_ENABLED ]: false,
				} );

			const { getByLabelText } = render(
				<SetupEnhancedMeasurementSwitch />,
				{
					registry,
				}
			);

			const switchControl = getByLabelText(
				'Enable enhanced measurement'
			);

			expect( switchControl ).not.toBeChecked();
		}
	);

	it( 'should toggle the switch on click', () => {
		const { getByLabelText } = render( <SetupEnhancedMeasurementSwitch />, {
			registry,
		} );

		const switchControl = getByLabelText( 'Enable enhanced measurement' );

		expect( switchControl ).toBeChecked();

		switchControl.click();

		expect( switchControl ).not.toBeChecked();
	} );

	describe.each( [
		[
			'accountID is changed',
			() => {
				registry.dispatch( MODULES_ANALYTICS ).setSettings( {
					accountID: '1001',
				} );
			},
		],
		[
			'propertyID is changed',
			() => {
				registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
					propertyID: '2001',
				} );
			},
		],
		[
			'webDataStreamID is changed',
			() => {
				registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
					webDataStreamID: '3001',
				} );
			},
		],
	] )( 'when the %s', ( _, changeSetting ) => {
		it( 'should revert the switch from off to on', () => {
			const { getByLabelText } = render(
				<SetupEnhancedMeasurementSwitch />,
				{
					registry,
				}
			);

			const switchControl = getByLabelText(
				'Enable enhanced measurement'
			);

			switchControl.click();

			expect( switchControl ).not.toBeChecked();

			act( changeSetting );

			expect( switchControl ).toBeChecked();
		} );

		it( 'should not toggle the switch from on to off', () => {
			const { getByLabelText } = render(
				<SetupEnhancedMeasurementSwitch />,
				{
					registry,
				}
			);

			const switchControl = getByLabelText(
				'Enable enhanced measurement'
			);

			expect( switchControl ).toBeChecked();

			act( changeSetting );

			expect( switchControl ).toBeChecked();
		} );
	} );
} );
