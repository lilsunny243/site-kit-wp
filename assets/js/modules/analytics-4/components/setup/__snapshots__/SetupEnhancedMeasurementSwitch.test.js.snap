// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupEnhancedMeasurementSwitch should render with the switch defaulting to the on position 1`] = `
<div>
  <div
    class="googlesitekit-analytics-enable-enhanced-measurement"
  >
    <div
      aria-checked="true"
      class="mdc-switch mdc-switch--checked"
      role="switch"
      tabindex="0"
    >
      <div
        class="mdc-switch__track"
      >
        Â 
      </div>
      <div
        class="mdc-switch__thumb-underlay"
      >
        <div
          class="mdc-switch__thumb"
        >
          <input
            class="mdc-switch__native-control"
            id="googlesitekit-switch-0"
            readonly=""
            role="switch"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <label
      class=""
      for="googlesitekit-switch-0"
    >
      Enable enhanced measurement
    </label>
    <p>
      Automatically measure interactions with your content (e.g. file downloads, form completions, video views). No extra code required. 
      <a
        aria-label="Learn more (opens in a new tab)"
        class="googlesitekit-cta-link googlesitekit-cta-link--external"
        href="https://support.google.com/analytics/answer/9216061?hl=en-US"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn more
      </a>
    </p>
  </div>
</div>
`;
